### ISSUE TITLE
Network feature does not use internal IP:s for cluster internal traffic

### ISSUE THREAD

**Issue by Troyhy** [mxschmitt/ui-driver-hetzner#44]
I have Rancher 2.2.4 cluster with latest ui driver and docker-machine-driver-hetzner_1.4.0_linux_amd64.tar.gz installed.

## Actual behaviour
After provisioning nodes, they get tag  `network=network-1` witch is correct. But listing nodes gives external ip <none> and internal IP is the actual external IP.
`kubectl get nodes -o wide`
```
NAME     STATUS   ROLES               AGE     VERSION   INTERNAL-IP      EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
ctrl-1   Ready    controlplane,etcd   4d23h   v1.14.1   95.216.0.1       <none>        Ubuntu 18.04.2 LTS   4.15.0-54-generic   docker://18.9.7
w-1      Ready    worker              3m41s   v1.14.1   95.216.0.2       <none>        Ubuntu 18.04.2 LTS   4.15.0-54-generic   docker://18.9.7
wrk-1    Ready    worker              7m22s   v1.14.1   95.216.0.3       <none>        Ubuntu 18.04.2 LTS   4.15.0-54-generic   docker://18.9.7
```

## Expected behaviour

External IP would be machine external IP and internal IP would be from network-1

**Comment by mxschmitt** [comment]
Hey @Troyhy,

I rechecked by using `iftop` for the the public and private interface. The Rancher related traffic, goes over the internal network: 
![image](https://user-images.githubusercontent.com/17984549/61710045-eefc1e80-ad50-11e9-903f-c9f99c990dcc.png)

But yes, the internal IP should be shown somewhere in the UI, maybe ask @vincent99, what do you think about this?

Best regards

Max

**Comment by mxschmitt** [comment]
I will close this issue here, because it seems to be a more generic Rancher related issue. This screenshot was made by using a machine created by using a Private Network of Digital Ocean.
![image](https://user-images.githubusercontent.com/17984549/62375244-22536000-b53e-11e9-9c90-5c8af4dc5d8f.png)

**Comment by vistalba** [comment]
Any update on this issue? Setup new rancher 2.2.7 and use internal network. All servers comming up fine... but all traffic inside the cluster still using the public network instead of the private one.

**Comment by mxschmitt** [comment]
@vistalba are you sure? Please check with `iftop`. Also you have to set the `server-url` in your global Rancher settings to the internal one.

**Comment by vistalba** [comment]
@mxschmitt I'm really sure that longhorn for example syncs the storage images over public... I can see it on grafana.

As you can see... nearly nothing is going trought interface `ens10` which is the private one:
![image](https://user-images.githubusercontent.com/12324196/63208452-821c4000-c0d4-11e9-9102-8a4cc9091ae4.png)

**Comment by mxschmitt** [comment]
I will investigate later and let you know.

**Comment by vistalba** [comment]
will try to get more information about this... but for the moment it looks like only rancher traffic is going trought private net. Will try to do some tcpdumps :)

**Comment by vistalba** [comment]
Okay... I can confirm... on the private interface `ens10` is no other traffic than `rancher <-> host` if I use `tcpdump -i ens10 -n -vvv`

I'm usinge weave as network. Maybe there is a issue....?

**Comment by mxschmitt** [comment]
I guess that this is a Rancher specific issue, I tried to reproduce it with Digitial Ocean Private Network a few weeks ago. On this cluster there is also the internal IP issue, so that the external one is used instead.

**Comment by vistalba** [comment]
Neverthless, does it work on DO? Does the internal traffic go over the private net or does it use the public too?

**Comment by mxschmitt** [comment]
Tbh, not tried. :/

**Comment by vistalba** [comment]
If I understand it right, we should set the internal-ip on Hetzner private interface as `--node-ip` in `KUBELET_KUBEADM_ARGS`. 
Or maybe you're driver should get the assigned ip from Hetzner API and pass it in a way to rancher?

**Comment by vistalba** [comment]
I just created a cluster manually and this works if I specify the `internal address` in the above screenshot:
![image](https://user-images.githubusercontent.com/12324196/63209809-04f9c680-c0e6-11e9-9346-2717fa1b99b5.png)

weave then starts: 
```
INFO: 2019/08/17 09:55:19.523589 Command line options: map[metrics-addr:0.0.0.0:6782 docker-api: no-dns:true conn-limit:100 datapath:datapath host-root:/host http-addr:127.0.0.1:6784 name:3e:a2:d6:10:f0:4b port:6783 db-prefix:/weavedb/weave-net expect-npc:true ipalloc-init:consensus=2 ipalloc-range:10.42.0.0/16 nickname:test-2]
INFO: 2019/08/17 09:55:19.523850 weave 2.5.0
INFO: 2019/08/17 09:55:19.645018 Bridge type is bridged_fastdp
INFO: 2019/08/17 09:55:19.645033 Communication between peers is unencrypted.
INFO: 2019/08/17 09:55:19.649538 Our name is 3e:a2:d6:10:f0:4b(test-2)
INFO: 2019/08/17 09:55:19.649577 Launch detected - using supplied peer list: [10.11.1.4 10.11.1.2]
```

and tcp dump also shows a lot of traffic in ens10. And not just between rancher and node .. also between the nodes.

```
11:56:59.788394 IP (tos 0x0, ttl 63, id 16933, offset 0, flags [DF], proto TCP (6), length 52)
    10.11.1.2.53292 > 10.11.1.4.2380: Flags [.], cksum 0x3766 (correct), seq 1777, ack 1011503, win 3878, options [nop,nop,TS val 599022462 ecr 3889185248], 
```

**Comment by vistalba** [comment]
@mxschmitt is it possible to update the driver so that it support this scenario?

As I understand there is no way to get the private ip from hetzner API. :-( 
You can just specify one by yourself if you attatch the network to a server. 
So maybe let people specify a ip-range inside the network they selected and the driver has to track wich one are already used? Or can you read it from the server interface itself?

**Comment by Troyhy** [comment]
I'm using hezner with Terraform and API has private IP available. 
 https://docs.hetzner.cloud/#servers-get-a-server
this is available after server is attached to network https://docs.hetzner.cloud/#server-actions-attach-a-server-to-a-network

**Comment by sthyx** [comment]
@mxschmitt Thanks for providing this great rancher ui (node) driver. 

Is there any news on this issue? Which module needs to be fixed:

1. hetzner machine driver (not giving the internal address)
or
2. rancher provisioning procedure (not utilizing the machine driver for setting the internal address)

Maybe we should open a ticket at the right place to take the next steps fixing the issue?

Unfortunately this is a show stopper for using rancher auto-scale in a hetzner environment because there is no secure traffic between nodes over the internal network possible.

**Comment by mxschmitt** [comment]
Hi @sthyx 
I was in contact with Vincent from the Rancher team. Docker machine has currently not the implementation to use it in combination with an Internal Network. So the issue is not on Rancher's side. Also it seems, that Docker is not maintaining Docker Machine anymore on a regular basis (60 open PRs). So more abandonware. Hmm no clue what I should do :/

**Comment by vitobotta** [comment]
I was trying the node driver again now and found this :(

**Comment by dilame** [comment]
PLEEEEEEASE

**Comment by vitobotta** [comment]
Just installed a new cluster and had forgotten about this. Docker Machine now has 937 open issues and 62 open pull requests... this isn't looking good :(

**Comment by vitobotta** [comment]
Hi @vistalba, if I use Weave doesn't that have encryption by default? So it shouldn't matter if the public network is used right?

**Comment by vistalba** [comment]
> Hi @vistalba, if I use Weave doesn't that have encryption by default? So it shouldn't matter if the public network is used right?

Yes, that should be the case and you can check it by `kubectl logs -n namespace weave weave-npc` if I remember correctly. Some of  the first lines after start weave indicates if it is encryptet or not.

**Comment by vitobotta** [comment]
@vistalba so I can rest assured that ALL the traffic is encrypted? For example I am going to use MySQL cluster, Redis cluster, Minio cluster and a Rails app that uses them. Can I assume that nobody can see the traffic between the nodes using weave? Thanks!

**Comment by vitobotta** [comment]
I see `Communication between peers via untrusted networks is encrypted.` in the logs

**Comment by vistalba** [comment]
> @vistalba so I can rest assured that ALL the traffic is encrypted? For example I am going to use MySQL cluster, Redis cluster, Minio cluster and a Rails app that uses them. Can I assume that nobody can see the traffic between the nodes using weave? Thanks!

Yes, if the weave overlay is used. So just for traffic from your docker workloads. Traffic between the VMs are not encrypted.

**Comment by vitobotta** [comment]
Wonderful, thanks a lot for the confirmation! I can sleep better now :D

**Comment by mxschmitt** [comment]
Hey folks (@Troyhy @vitobotta @sthyx @vistalba @dilame)

Now the feature finally went into the project. See PR #83. I'm closing it for now, feel free to reopen or ask questions.

**Comment by vitobotta** [comment]
@mxschmitt This is awesome!!! Just yesterday I rebuilt my cluster with Pulumi and used Weave again due to the encryption. I may do it again with private network and Canal since I think performance would benefit. Is all the traffic definitely going through the private network - as long as we specify the interface in the Canal settings? Does the new driver ensure that nodes get private IPs? Or is there something we need to do to make sure all traffic between the nodes is private?

Thanks A LOT!

**Comment by vitobotta** [comment]
Oh I just saw the pictures in the PR properly, I can see the private IP assigned. Wonderful! 

So is it safe now to just use Canal with this? Are there still any risks that some Hetzner customer might intercept traffic or something like that?

**Comment by CuBiC3D** [comment]
> So is it safe now to just use Canal with this? Are there still any risks that some Hetzner customer might intercept traffic or something like that?

See https://wiki.hetzner.de/index.php/CloudServer/en#Is_traffic_inside_Hetzner_Cloud_Networks_encrypted.3F
Hetzner does not encrypt any Traffic. It is Isolated by encapsulation, but I assume it is very well interceptable between datacenters since there are no dedicated lines between them in different countries. I can not find any further information on that, you would need to ask the support how the situation looks like for a single datacenter if the traffic presumably not leaving Hetzner premises.

@mxschmitt Thank you very much for following up on this issue! I'll start a test deploy soon.

**Comment by vitobotta** [comment]
What is private then? What does isolated by encapsulation mean? My networking knowledge is poor unfortunately,

**Comment by vitobotta** [comment]
I just remembered that I asked support about private networks a while ago! This was their answer:

`It is not possible for another customer to intercept/see your internal network traffic. You dont have to encrypt your communication.`

Do you believe it?

**Comment by CuBiC3D** [comment]
That's a bit offtopic on the issue and only assumptions about Hetzner's infrastructure:

Private Network mens that the subnet belongs to you and is not routed on the Internet since its not unique. You can expect to have no other traffic on this subnet except your servers. This traffic is not forwarded to any other customer's servers - that should be correct so far (again, I can't check that).

But, private networks are possible between different datacenters like Falkenstein and Helsinki and have to be forwarded through the internet. The packets are encapsulated and carry an ID which identifies your private network inside Hetzner Cloud but every forwarder on the internet sees the packet unencrypted (again, that's an assumption, since they said there's no encryption in place).

So basically, the packets are not encrypted on transport and could be intercepted by the nodes involved to transport your packet - other customers are not involved in transportation. If you are located in Germany this could also be a problem in terms of DSGVO if your packets carry person related data.

**Comment by vitobotta** [comment]
What if I use the private network only within a single DC?

**Comment by CuBiC3D** [comment]
> What if I use the private network only within a single DC?

Then the forwarders should be Hetzner routers only, but no way for me to know that (and leaks can always be possible by misconfiguration of routers). Ask the support about that.

But if basically depends on your trust in the company. You trust them to store your data in the cloud but not on transportation -> encrypt it. Trust them on both -> do not encrypt if it doesn't leave the DC. Depends also on the data and your customers.

**Comment by vitobotta** [comment]
Well I have to trust the company... they have anyway access to my data if they want

**Comment by mxschmitt** [comment]
I've added a useful information section [to the README](https://github.com/mxschmitt/ui-driver-hetzner#useful-resources) with tips & tricks regarding the usage with the private network feature. Feel free to contribute there or ping me if anything is wrong. Thanks!

**Comment by MrBenFTW** [comment]
Hi, I believe this is similar to what I am trying to do so is it possible you could tell me how you managed to fix this?

I have a /28 subnet at Hetzner, I have created a bridge and given my docker container an IP on that subnet.

Externally, I can access the container using the IP I gave it. However, if `curl api.ipify.org` then the ip returned is the IP of my dedicated server and not the container. 

This wouldn't be too much of a problem except that I have a mail server in a container and its reverse dns record is incorrect. I've been trying to fix this for weeks but I keep hitting a dead end
